Refactor Terminal Module

arafatkatze opened the pull request:
## Description

This PR refactors the terminal command execution system to eliminate code duplication and improve maintainability. The previous implementation had ~850 lines of duplicated orchestration logic between `VscodeCommandExecutor.ts` and `StandaloneCommandExecutor.ts`. This refactoring extracts the shared logic into a unified architecture.

### Key Changes:

1. **Created `ITerminalProcess` interface** (`src/integrations/terminal/ITerminalProcess.ts`)
   - Defines a common interface for terminal processes with events: `line`, `completed`, `error`, `no_shell_integration`
   - Both `VscodeTerminalProcess` and `StandaloneTerminalProcess` now implement this interface

2. **Created `CommandOrchestrator`** (`src/integrations/terminal/CommandOrchestrator.ts`)
   - Extracted ~300 lines of shared orchestration logic (buffering, user interaction, "Proceed While Running", timeout handling)
   - Used by the unified `CommandExecutor` for all terminal modes

3. **Unified `CommandExecutor`** (`src/integrations/terminal/CommandExecutor.ts`)
   - Single executor class (~260 lines) that handles both VSCode terminal and standalone/background modes
   - Routes subagent commands to `StandaloneTerminalManager` (hidden execution)
   - Delegates orchestration to `CommandOrchestrator`

4. **Deleted redundant executor classes**
   - Removed `src/integrations/terminal/executors/VscodeCommandExecutor.ts` (~370 lines)
   - Removed `src/integrations/terminal/executors/StandaloneCommandExecutor.ts` (~480 lines)

5. **Updated Task class** (`src/core/task/index.ts`)
   - Now uses the unified `CommandExecutor` with proper configuration
   - Terminal manager selection happens at Task construction time

### Code Reduction:
- **Before:** ~850 lines of duplicated code across two executor classes
- **After:** ~560 lines total (unified executor + shared orchestrator)
- **Reduction:** ~34% less code, zero duplication

## Test Procedure
1. Run `npm run compile` - should pass with no TypeScript errors ✅
2. Run `npm test` - all existing tests should pass
3. Test VSCode extension terminal commands work as expected
4. Test CLI mode terminal commands work as expected
5. Test subagent commands (cline CLI) route to hidden terminal
6. Test "Proceed While Running" button functionality
7. Test command cancellation in background mode

## Type of Change
- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] ♻️ Refactor Changes
- [ ] 💅 Cosmetic Changes
- [ ] 📚 Documentation update
- [ ] 🏃 Workflow Changes

## Pre-flight Checklist
- [x] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
- [ ] Tests are passing (`npm test`) and code is formatted and linted (`npm run format && npm run lint`)
- [ ] I have created a changeset using `npm run changeset`

### Screenshots

N/A - This is a backend refactoring with no UI changes. The terminal command execution behavior should be identical to before.

### Additional Notes

**Architecture After Refactoring:**
```
ITerminalProcess (shared interface)
    ├── VscodeTerminalProcess implements ITerminalProcess
    └── StandaloneTerminalProcess implements ITerminalProcess

CommandExecutor (unified, ~260 lines)
    └── orchestrateCommandExecution() (shared, ~300 lines)
        └── ITerminalManager.runCommand() (polymorphic)
            ├── VscodeTerminalManager → VscodeTerminalProcess
            └── StandaloneTerminalManager → StandaloneTerminalProcess
```

**Parity Notes:**
- VSCode terminal mode and background exec mode now share identical orchestration logic
- Cancellation is only available in standalone/background mode (VSCode terminals can't be programmatically terminated)
- Subagent commands always route to StandaloneTerminalManager regardless of configured mode




### Additional Notes

<!-- Add any additional notes for reviewers -->

<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Refactor terminal command execution by introducing a unified architecture with `ITerminalProcess`, `CommandOrchestrator`, and `CommandExecutor`, reducing code duplication and improving maintainability.
> 
>   - **Behavior**:
>     - Introduces `ITerminalProcess` interface in `ITerminalProcess.ts` for terminal processes with events like `line`, `completed`, `error`, `no_shell_integration`.
>     - Implements `ITerminalProcess` in `VscodeTerminalProcess` and `StandaloneTerminalProcess`.
>     - Creates `CommandOrchestrator` in `CommandOrchestrator.ts` for shared orchestration logic.
>     - Unified `CommandExecutor` in `CommandExecutor.ts` for handling both VSCode and standalone modes.
>     - Routes subagent commands to `StandaloneTerminalManager`.
>   - **Code Reduction**:
>     - Deletes `VscodeCommandExecutor.ts` and `StandaloneCommandExecutor.ts`, reducing ~850 lines of duplicated code to ~560 lines.
>   - **Task Class**:
>     - Updates `Task` class in `task/index.ts` to use the unified `CommandExecutor`.
>     - Terminal manager selection occurs at Task construction.
>   - **Misc**:
>     - Moves `get-latest-output.ts` to `hosts/vscode/terminal/`.
>     - Removes shell integration warning logic from `controller/index.ts`.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for 3576f695768977486672fc8c7ae43cc8737bc0ec. You can [customize](https://app.ellipsis.dev/cline/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

sjf commented on src/core/task/index.ts:
```diff
@@ -498,6 +486,45 @@ export class Task {
+			terminalManager: this.terminalManager,
+			taskId: this.taskId,
+			ulid: this.ulid,
+			shouldShowBackgroundTerminalSuggestion: () => this.controller.shouldShowBackgroundTerminalSuggestion(),
```
Can you move `shouldShowBackgroundTerminalSuggestion` out of the controller into `CommandExecutor`; it's not used anywhere else and its not using anything specific to the controller.

arafatkatze commented:
Done! I've moved `shouldShowBackgroundTerminalSuggestion` out of the Controller and into `CommandExecutor` where it's actually used.

The change includes:
- Added `shellIntegrationWarningTracker` to CommandExecutor
- Added `shouldShowBackgroundTerminalSuggestion()` method to CommandExecutor  
- Removed the callback from `CommandExecutorConfig`
- Removed the tracker and method from Controller

This improves encapsulation since the warning tracking logic now lives with the code that uses it.

arafatkatze commented:
Good point! Removed the `ICommandExecutor` interface since there's only one implementation. The class now just exports directly without the extra indirection.

arafatkatze commented on src/core/mentions/index.ts:
```diff
@@ -12,6 +11,7 @@ import fs from "fs/promises"
 import { isBinaryFile } from "isbinaryfile"
 import * as path from "path"
 import { HostProvider } from "@/hosts/host-provider"
+import { getLatestTerminalOutput } from "@/hosts/vscode/terminal/get-latest-output"
```
Good catch! This import has been here for the `@terminal` mention feature (lets users include terminal output in their messages). It predates the terminal refactoring work in this PR.

I can either:

1. __Add a dynamic import with host check__ - so it only loads on VSCode and returns a fallback message on other hosts
2. __Leave it for a follow-up PR__ - since it's pre-existing and not introduced by this PR

Let me know which you'd prefer. Happy to fix it here or track it separately.

arafatkatze replied:
Yeah I can't think of a great way either to be honest. I mean one way might be to totally get rid of the feature because I doubt if anyone uses it.

arafatkatze reviewed (commented):
Looking at this more closely, I noticed `ICommandExecutor.ts` could be consolidated into `CommandExecutor.ts`:

- `CommandExecutorCallbacks` and `ActiveBackgroundCommand` are only used by `CommandExecutor.ts` and `CommandOrchestrator.ts`
- `AskResponse` is only used internally by `CommandExecutorCallbacks`
- The `CommandExecutorConfig` in `ICommandExecutor.ts` appears to be unused (there's a different, more complete version in `CommandExecutor.ts`)

Since this PR is about reducing redundancy, consider moving these types into `CommandExecutor.ts` and deleting `ICommandExecutor.ts`. Then update the imports in `CommandOrchestrator.ts` and `index.ts` accordingly.

This would remove one file and eliminate the confusing duplicate `CommandExecutorConfig` definition.
